doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+The polymorphic versions of those functions, if exist, are usually provided by third-party libraries like cats.
+
+For Plutus Tx, even with `nonstrict`, it is still useful to provide monomorphic versions of certain functions in the standard library, especially since they can be noticeably cheaper than the polymorphic versions.
+This will be discussed in the next section.
+
+### Monomorphizing Plutus Tx Standard Library
+
+Strict languages do not lend themselves to code reuse, as there is a high cost to use abstractions such as higher-order functions and parametric polymorphism.
+Polymorhpic functions taking `Foldable` or similar constraints, implemented in terms of higher-order functions like `foldr`, are rare in strict languages.
+
+For Plutus Tx, there's another reason why monomorphization of the standard library is beneficial: multiple method typeclasses are expensive.
+
+Single method typeclasses are cheap because they have simpler representations in GHC Core - a newtype that can be compiled away.
+On the other hand, multiple method typeclasses exist in GHC Core as actual dictionaries and need to be compiled into product types in PIR.
+
+This makes monomorphization the only choice in many cases. Consider `foldr` and `foldMap`.
+They can be implemented in terms of each other, so technically, only one of them needs to be in the `Foldable` class.
+But certain types have more efficient direct implementations of `foldr`; the same for `foldMap`.
+No matter which of them we move out of `Foldable`, it won't work optimally for certain types.
+Keeping both in `Foldable` will cause `Foldable` to be a multiple method typeclass, which is expensive in and of itself, as explained above.
+We may as well do away with `PlutusTx.Foldable`, and instead simply provide monomorphic versions of `Foldable`-related functions for each common type.
+
+### Turning off GHC's pre-inliner
+
+As said before, one reason why the semantics of Plutus Tx is difficult to predict is because we run GHC's pre-inliner, which may perform unconditionally inlining.
+Since GHC 9.0, GHC provides [an option](https://hackage.haskell.org/package/ghc-9.6.1/docs/GHC-Core-Opt-Simplify-Env.html#v:sm_pre_inline) to turn it off when running the simplifier.
+By turning it off, GHC (hopefully) won't inline anything at all, which means it won't change the strictness of Plutus Tx programs.
+
+At this time we can't simply turn the pre-inliner off, because doing so breaks the Plutus Tx compiler.
+For details, see [this note](https://github.com/input-output-hk/plutus/blob/dc57fe1b8497835ef2a7794c017c5a50c0b7e647/plutus-tx-plugin/src/PlutusTx/Plugin.hs#L123).
+The problem described in the note should, however, be resolved in a different way, as explained in the note.
+Once done, we can go ahead and turn it off.
+
+### Performing GHC-like Inlining for Functions with the INLINE Pragma
```

Comment:
Can this section be moved right under the "Using `nonstrict` to Make Arguments Non-strict" section? And perhaps even rename this section to "Using the INLINE pragma to define non-strict functions" or something like that?

Reply:
Sounds good.